CI: Add Riscv bare metal job

sedited opened the pull request:
This adds a CI job for building the static consensus library and linking it to an executable. It uses newlib-cygwin as a C library for the final linking step. This ensure compatibility with this target going forward and can serve as a starting point for enabling bare metal builds for the entire kernel library. This would have also caught the error fixed in #31365.

sedited committed:
build: Add option for building for bare metal envs

A bare metal build is now supported by setting CMAKE_SYSTEM_NAME=Generic

Skip the platform-dependent feature checks, such as threads and atomics,
which are typically not available on bare metal. Also only make the
boost headers mandatory if they exist for the target.

Also exclude aes from the crypto library, since that requires the secure
allocator, which uses more advance features, such as mutexes.

sedited committed:
Add CI job for producing a static bare metal binary

TheCharlatan commented:
> Interesting. Do you think this is possible at all, given that the kernel library links leveldb, which I'd presume is not bare-metal ready?

Yes, definitely not ready. I think the main hurdle is the background compaction, which I am not sure how to tackle. Maybe we'll find a solution for it eventually though, either by patching it, or allowing the user to bring their own utxos.

> So I guess this mostly serves as a check that users can ship their own-brewed libbitcoinconsensus (or so, or subset of it)?

Yes, that is the goal for now.

laanwj commented on ci/test/03_test_script.sh:
```diff
@@ -132,6 +132,44 @@ bash -c "${PRINT_CCACHE_STATISTICS}"
+        src/secp256k1/lib/libsecp256k1.a \
+        /opt/riscv-ilp32/riscv32-unknown-elf/lib/libstdc++.a \
+        /riscv/newlib/build/riscv32-unknown-elf/newlib/libc.a \
+        /riscv/newlib/build/riscv32-unknown-elf/newlib/libm.a \
```
Any specific reason to use the intermediate build and not the installed libraries from `/opt/newlib`, here?

laanwj commented on ci/test/01_base_install.sh:
```diff
@@ -84,6 +84,29 @@ if [[ ${USE_MEMORY_SANITIZER} == "true" ]]; then
+    cd /riscv/newlib
+    mkdir build && cd build
+    ../configure \
+        --target=riscv32-unknown-elf --disable-newlib-io-float --enable-newlib-io-long-long --enable-newlib-io-long-double --with-arch=rv32gc --with-abi=ilp32 --disable-shared --disable-multilib\
```
Are you sure enabling i/o for long-double is needed? i don't believe we use this type anywhere.

laanwj commented:
> It uses newlib-cygwin as a C library for the final linking step. 

Mentioning this because i had to look it up to be sure: [newlib-cygwin](https://en.wikipedia.org/wiki/Newlib) has nothing to do with Windows whatsoever. It's simply a minimalist libc.

> I think the main hurdle is the background compaction, which I am not sure how to tackle.

Could be a periodic foreground task, if threads aren't available? But yes, this would imply patching leveldb, there is no such API right now.

TheCharlatan commented:
Updated 4910ae5d0d9954b8b811d7c52e3539979ea46016 -> f27760012b3e236933ea5d1b39a6ba74884df1a2 ([bare_metal_support_0](https://github.com/TheCharlatan/bitcoin/tree/bare_metal_support_0) -> [bare_metal_support_1](https://github.com/TheCharlatan/bitcoin/tree/bare_metal_support_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/bare_metal_support_0..bare_metal_support_1))

* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r1871400205), removing sources after install step.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r1871396146), wrapped cd step into a subshell
* Addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r1871190757), use the installed lib
* Addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r1871194413), removing a bunch of unneeded/superfluous flags
* Added a start section to the binary for the global pointer and returning an exit code. Though not needed, since we are only checking that it links, I feel like this makes the example a bit clearer.

laanwj commented on ci/test/03_test_script.sh:
```diff
@@ -132,6 +132,61 @@ bash -c "${PRINT_CCACHE_STATISTICS}"
+
+              call main
+
+              # Put Exit2 system call number into the a7 register
```
concept ACK on making it actually runnable
i was confused for a moment here, as to what syscall numbers mean for bare-metal
maybe add "Linux" to the comment

laanwj commented:
Concept ACK, this looks like the minimum required to sanity check that libconsensus.a can be compiled for, and linked for bare-metal RISC-V.
It is not enough to test that it actually works, but maybe that's out of scope for this project. At the least it is for this PR.

hebasto commented:
Concept ACK.

hebasto commented on CMakeLists.txt:
```diff
@@ -372,8 +374,10 @@ if(BUILD_FUZZ_BINARY)
   )
 endif()
 
+if(NOT CMAKE_SYSTEM_NAME STREQUAL "Generic")
```
I imagine that expanding this build scenario might enable building `boost` and `libevent` in depends, no?

If so, the Boost check should have another option to be disabled.

Additionally, mind considering this comment: https://github.com/bitcoin/bitcoin/blob/1172bc4157eefe80d1aaf0b56459857ec651e535/cmake/module/AddBoostIfNeeded.cmake#L7-L8

which, in turn, was based on this legacy code:https://github.com/bitcoin/bitcoin/blob/32efe850438ef22e2de39e562af557872a402c31/configure.ac#L1247-L1251

fanquake commented:
I think the docs might be incomplete / outdated? See https://github.com/bminor/newlib/blob/b39b510c1ce68757e79410585262ca2cd48da839/newlib/libc/include/sys/unistd.h#L287.

TheCharlatan commented:
Rebased f27760012b3e236933ea5d1b39a6ba74884df1a2 -> ec7c86c732c995d2c38e2a3f93ad55a451a8cf81 ([bare_metal_support_1](https://github.com/TheCharlatan/bitcoin/tree/bare_metal_support_1) -> [bare_metal_support_2](https://github.com/TheCharlatan/bitcoin/tree/bare_metal_support_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/bare_metal_support_1..bare_metal_support_2))

* Fixed a bunch of silent and proper merge conflicts.

sedited committed:
Add CI job for riscv bare metal

sedited commented:
Rebased 774042bb0eecee783ab04e0cb53f762b71984e77 -> 0de648ca1e71cdb76883dc2e31f4c7e630528d27 ([bare_metal_support_6](https://github.com/sedited/bitcoin/tree/bare_metal_support_6) -> [bare_metal_support_7](https://github.com/sedited/bitcoin/tree/bare_metal_support_7), [compare](https://github.com/sedited/bitcoin/compare/bare_metal_support_6..bare_metal_support_7))

* Added some more workarounds after the recent changes to the aes code included the secure allocator, which pulls in the lockedpool, which in turn does not make sense in its current form for bare metal targets.

sedited commented:
Updated 0de648ca1e71cdb76883dc2e31f4c7e630528d27 -> 50fc04c18cb226bf7545698bcabd1f0028f253cd ([bare_metal_support_7](https://github.com/sedited/bitcoin/tree/bare_metal_support_7) -> [bare_metal_support_8](https://github.com/sedited/bitcoin/tree/bare_metal_support_8), [compare](https://github.com/sedited/bitcoin/compare/bare_metal_support_7..bare_metal_support_8))

* Fixed conflict with #33974

fanquake commented on .github/workflows/ci.yml:
```diff
@@ -537,6 +537,12 @@ jobs:
             file-env: './ci/test/00_setup_env_native_msan.sh'
 
+          - name: 'riscv32 bare metal, static libbitcoin_consensus'
+            cirrus-runner: 'ghcr.io/cirruslabs/ubuntu-runner-amd64:24.04-lg'
```
Can switch this to:
```yml
            warp-runner: 'warp-ubuntu-latest-x64-16x'
            fallback-runner: 'ubuntu-latest'
```

fanquake commented on ci/test/00_setup_env_riscv_bare_cross.sh:
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2019-present The Bitcoin Core developers
```
```bash
# Copyright (c) The Bitcoin Core developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or https://opensource.org/license/mit.
```

fanquake commented on ci/test/00_setup_env_riscv_bare_cross.sh:
```diff
@@ -0,0 +1,37 @@
+export CONTAINER_NAME=ci_native_riscv_bare
+
+export GOAL="bitcoin_consensus bitcoin_crypto secp256k1"
+export CI_IMAGE_NAME_TAG="docker.io/ubuntu:24.04"
```
```suggestion
export CI_IMAGE_NAME_TAG="mirror.gcr.io/ubuntu:26.04"
```

fanquake commented on ci/test/01_base_install.sh:
```diff
@@ -84,6 +84,29 @@ if [[ -n "${USE_INSTRUMENTED_LIBCPP}" ]]; then
 fi
 
+if [[ ${BARE_METAL_RISCV} == "true" ]]; then
+    ${CI_RETRY_EXE} git clone --depth=1 https://github.com/riscv-collab/riscv-gnu-toolchain -b 2024.11.22 /riscv/gcc
```
> 2024.11.22

Should this point to any particular release (can't tell from upstream tags, which are all nightly, what should be used)?

sedited replied:
This basically just needs a compatible host. Looks like you need to look at the submodule information to actually find out what is being versioned? I don't think it matters too much as long as everything links fine.

fanquake commented on ci/test/03_test_script.sh:
```diff
@@ -167,6 +167,62 @@ if [ -n "${CI_LIMIT_STACK_SIZE}" ]; then
   ulimit -s 512
 fi
 
+if [[ ${BARE_METAL_RISCV} == "true" ]]; then
```
This chunk is a lot to have inlined here. Could split it into it's own script?

sedited replied:
Could do that, but not sure where to put.

maflcko replied:
You could put it in `"${BASE_ROOT_DIR}/ci/test/` like some other ci test snippets.

fanquake reviewed (commented):
I think this is something we should be testing, and would like to see it move forward.

willcl-ark commented on ci/test/03_test_script.sh:
```diff
@@ -167,6 +167,62 @@ if [ -n "${CI_LIMIT_STACK_SIZE}" ]; then
+        /opt/newlib/riscv32-unknown-elf/lib/libc.a \
+        /opt/newlib/riscv32-unknown-elf/lib/libm.a \
+        /opt/riscv-ilp32/lib/gcc/riscv32-unknown-elf/14.2.0/libgcc.a \
+        -o test.elf
```
As an interim, how about force-linking the whole archive, while test only has ScriptErrorString. Otherwise the linker can ignore everything else no?

```suggestion
    /opt/riscv-ilp32/bin/riscv32-unknown-elf-g++ -g -std=c++20 -march=rv32gc -mabi=ilp32 \
        -nostdlib \
        /opt/riscv-ilp32/lib/gcc/riscv32-unknown-elf/14.2.0/crtbegin.o \
        test.o \
        start.o \
        syscalls.o \
         -Wl,--whole-archive \
        "${BASE_BUILD_DIR}"/lib/libbitcoin_consensus.a \
        "${BASE_BUILD_DIR}"/lib/libbitcoin_crypto.a \
        "${BASE_BUILD_DIR}"/src/secp256k1/lib/libsecp256k1.a \
         -Wl,--no-whole-archive \
        /opt/riscv-ilp32/riscv32-unknown-elf/lib/libstdc++.a \
        /opt/newlib/riscv32-unknown-elf/lib/libc.a \
        /opt/newlib/riscv32-unknown-elf/lib/libm.a \
        /opt/riscv-ilp32/lib/gcc/riscv32-unknown-elf/14.2.0/libgcc.a \
        -o test.elf
```

sedited commented:
Rebased and updated 50fc04c18cb226bf7545698bcabd1f0028f253cd -> 0558992d33f336790721920dd57d17441a6c9cea ([bare_metal_support_8](https://github.com/sedited/bitcoin/tree/bare_metal_support_8) -> [bare_metal_support_9](https://github.com/sedited/bitcoin/tree/bare_metal_support_9), [compare](https://github.com/sedited/bitcoin/compare/bare_metal_support_8..bare_metal_support_9))

* Updated to newest riscv toolchain tag and modernized the script a bit, allowing to remove the external newlib and use the version shipped with the toolchain repository.
* Addressed @willcl-ark's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r3429599341), added `whole-archive` flags.
* Addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r3427218486), switched to warp runners.
* Addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r3427222529), corrected copyright notice.
* Addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r3427225087), migrated to ubuntu 26.04

willcl-ark commented on ci/test/03_test_script.sh:
```diff
@@ -167,6 +167,72 @@ if [ -n "${CI_LIMIT_STACK_SIZE}" ]; then
 fi
 
+if [[ ${BARE_METAL_RISCV} == "true" ]]; then
+    echo -e "#include <script/script_error.h>\n int main() { return ScriptErrorString(ScriptError_t::SCRIPT_ERR_UNKNOWN_ERROR).size() > 0; }" > test.cpp
```
In bff80272432f8393651ce53cd0be9a5123490b41

Yeah, I agree with fanquake/marco that this assembly/syscall stub/linking section is not really test orchestration, so keeping it inline in `03_test_script.sh` makes that shared script harder to read.

As suggested previously, could we move the generated `test.cpp`, `start.s`, `syscalls.c`, and final link command into a dedicated helper, e.g. `${BASE_ROOT_DIR}/ci/test/link_riscv_bare_consensus.sh`, and have `03_test_script.sh` just call that helper when `BARE_METAL_RISCV=true`?

willcl-ark replied:
I see this was actioned before I could hit send! :)

sedited commented:
Updated 0558992d33f336790721920dd57d17441a6c9cea -> fcbf52fdd1b91298ab81a7dde3063554826f32a0 ([bare_metal_support_9](https://github.com/sedited/bitcoin/tree/bare_metal_support_9) -> [bare_metal_support_10](https://github.com/sedited/bitcoin/tree/bare_metal_support_10), [compare](https://github.com/sedited/bitcoin/compare/bare_metal_support_9..bare_metal_support_10))

* Addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/31425#discussion_r3427286100), split linking steps into their own script.
* Disable gdb for the toolchain, which slims down the build a bit.

willcl-ark reviewed (commented):
Approach ACK

this CI job gives an early warning if `libbitcoin_consensus.a`, `libbitcoin_crypto.a`, or `libsecp256k1.a` start depending on hosted OS assumptions that break a `riscv32-unknown-elf` bare-metal-style link. although using `generic` I think will actually help catch a _class_ of errors which may apply to other bare metal targets too.

it seems therefore like a useful minimal guard for keeping the consensus-related static archive linkable in a freestanding environment, and @sedited already highlighted one bug which would have been caught if we'd had this job.

however it is currently only a compile/link smoke test. it doesn't run the ELF, boot on hardware, exercise script verification, or prove broader bare-metal support. (not goals of this PR, either). the handwritten `_start`, syscall stubs, and link recipe are also currently embedded in the generic CI script.

follow ups which I can think of as being interesting might be to link a small real `bitcoinconsensus_verify_script` caller, and run the ELF under QEMU.

willcl-ark commented on .github/workflows/ci.yml:
```diff
@@ -523,6 +523,12 @@ jobs:
             file-env: './ci/test/00_setup_env_native_msan.sh'
 
+          - name: 'riscv32 bare metal, static libbitcoin_consensus'
+            warp-runner: 'warp-ubuntu-latest-x64-16x'
```
I think it would be good to see if we can downgrade this to 8x after it's been merged here. It's only 10 minutes long, and will likely get even shorter when hitting more caches.

Generally, it's better (cheaper) for us to use a slower CI instance, if we can tolerate the longer runtimes. As this one appears to be pretty quick (~ 10 minutes) it could make sense to downgrade to 8x in the future IMO.

fanquake replied:
Can take a look in #35640.

willcl-ark reviewed (approved):
ACK fcbf52fdd1b91298ab81a7dde3063554826f32a0

Thjis LGTM now.  I'm sure we will find more improvements to make to it, but it already serves a useful purpose, and looks to be implemented correctly according to it's current intended purpose.

sedited commented:
Rebased fcbf52fdd1b91298ab81a7dde3063554826f32a0 -> b36730a3efc85c4acdb93f06910e37151b8f2cf9 ([bare_metal_support_10](https://github.com/sedited/bitcoin/tree/bare_metal_support_10) -> [bare_metal_support_11](https://github.com/sedited/bitcoin/tree/bare_metal_support_11), [compare](https://github.com/sedited/bitcoin/compare/bare_metal_support_10..bare_metal_support_11))

* Fixed conflict with #34411

willcl-ark reviewed (approved):
reACK b36730a3efc85c4acdb93f06910e37151b8f2cf9

Checked range-diff only

```
$ git range-diff fcbf52fdd1b...b36730a3efc

 -:  ----------- > 87:  703a671fbc2 fuzz: compact coins view db during fuzzing
 1:  29af6df7e91 ! 88:  a9a1d92a1da build: Add option for building for bare metal envs
    @@ CMakeLists.txt: endif()
      add_boost_if_needed()
     +endif()

    - if(BUILD_DAEMON OR BUILD_GUI OR BUILD_CLI OR BUILD_TESTS OR BUILD_BENCH OR BUILD_FUZZ_BINARY)
    -   find_package(Libevent 2.1.8 MODULE REQUIRED)
    + if(ENABLE_WALLET)
    +   if(VCPKG_TARGET_TRIPLET)

      ## cmake/introspection.cmake ##
     @@ cmake/introspection.cmake: if(NOT WIN32)
    @@ src/CMakeLists.txt: target_link_libraries(bitcoin_node
          univalue
     -    Boost::headers
     +    $<TARGET_NAME_IF_EXISTS:Boost::headers>
    -     $<TARGET_NAME_IF_EXISTS:libevent::core>
    -     $<TARGET_NAME_IF_EXISTS:libevent::extra>
    -     $<TARGET_NAME_IF_EXISTS:libevent::pthreads>
    +     $<TARGET_NAME_IF_EXISTS:USDT::headers>
    + )
    + if(WITH_EMBEDDED_ASMAP)

      ## src/crypto/CMakeLists.txt ##
     @@
 2:  bfcbc99953e = 89:  bfdbf513f69 Add CI job for producing a static bare metal binary
 3:  fcbf52fdd1b = 90:  b36730a3efc  Add CI job for riscv bare metal
```

fanquake commented:
Guix Build:
```bash
6be5510ac97201592d603a27c9a505de38c458114788d70dd661fe8fd3d6e6f8  guix-build-b36730a3efc8/output/aarch64-linux-gnu/SHA256SUMS.part
5e45521428d5238e033a7d389fff528cbca87ab90896064d750d025640703c69  guix-build-b36730a3efc8/output/aarch64-linux-gnu/bitcoin-b36730a3efc8-aarch64-linux-gnu-debug.tar.gz
f6b3984af843fe120e66e9666b4b8a089ce2b840b9a2357d8b44e2a8eda4410b  guix-build-b36730a3efc8/output/aarch64-linux-gnu/bitcoin-b36730a3efc8-aarch64-linux-gnu.tar.gz
4c8e7b5f32293878baa2182e64195ff0c0dfb17241d0b1f08e1d4f4a39f4adfd  guix-build-b36730a3efc8/output/arm-linux-gnueabihf/SHA256SUMS.part
44f106cccc1d7c140109476c94d60742e6cd7bae5add7524d22447f7cc9721de  guix-build-b36730a3efc8/output/arm-linux-gnueabihf/bitcoin-b36730a3efc8-arm-linux-gnueabihf-debug.tar.gz
77a191511717860710465402eaf76682b9febda71877faf4d808017dfe610825  guix-build-b36730a3efc8/output/arm-linux-gnueabihf/bitcoin-b36730a3efc8-arm-linux-gnueabihf.tar.gz
553c96e9f0b89fb1aa36a92f1b7358756afb18f74ef228bc48d843dcc74955f2  guix-build-b36730a3efc8/output/arm64-apple-darwin/SHA256SUMS.part
c6722e53e2297a0e4ba278f0a9c2db7f5aac46cd08a8c306dab9a2c717aed357  guix-build-b36730a3efc8/output/arm64-apple-darwin/bitcoin-b36730a3efc8-arm64-apple-darwin-codesigning.tar.gz
d812ae4608f5a82cf426a2092e368d0a65414e854c29124f49a86cf914eff94d  guix-build-b36730a3efc8/output/arm64-apple-darwin/bitcoin-b36730a3efc8-arm64-apple-darwin-unsigned.tar.gz
c5f54764eae63cfbf0eb029ba7c8c82073cf8bb9fd577bf780f701d09beccc1a  guix-build-b36730a3efc8/output/arm64-apple-darwin/bitcoin-b36730a3efc8-arm64-apple-darwin-unsigned.zip
1266fafab396a1fc3395ea4767c8e36225b05537f34ba759936677f07dced0f3  guix-build-b36730a3efc8/output/dist-archive/bitcoin-b36730a3efc8.tar.gz
2989ef62e3ea922f4d20b2cd819ea6d5a92fee9df44e0625aa3fee8de0126688  guix-build-b36730a3efc8/output/powerpc64-linux-gnu/SHA256SUMS.part
2839d7dd674212a55e3c4607c371233b11a7532e080c2f7356732aa629ecda7a  guix-build-b36730a3efc8/output/powerpc64-linux-gnu/bitcoin-b36730a3efc8-powerpc64-linux-gnu-debug.tar.gz
474fe25da36be086ea228a4d5d7684f49fdb0a6227bc10d49a6adb796921316d  guix-build-b36730a3efc8/output/powerpc64-linux-gnu/bitcoin-b36730a3efc8-powerpc64-linux-gnu.tar.gz
590191adf09b41365e80d28e9b1ebee7961154cb733767aad2ca5d38ad3073b8  guix-build-b36730a3efc8/output/riscv64-linux-gnu/SHA256SUMS.part
61dacc898e24b439d747840b3551b76731809529ae426caad9e0427eb86db641  guix-build-b36730a3efc8/output/riscv64-linux-gnu/bitcoin-b36730a3efc8-riscv64-linux-gnu-debug.tar.gz
af959fce01450358ec25da5400d1aaf004fa43a7ba5795833036beeb6ee0a935  guix-build-b36730a3efc8/output/riscv64-linux-gnu/bitcoin-b36730a3efc8-riscv64-linux-gnu.tar.gz
31925acd25f9e26e29ecdafdf1b4866e457aa08d64154aa5a82bbca1e49beaba  guix-build-b36730a3efc8/output/x86_64-apple-darwin/SHA256SUMS.part
b6c36c78dac4145dc37e2318913b778e26a05e84e3549632daaa28dfc11e270c  guix-build-b36730a3efc8/output/x86_64-apple-darwin/bitcoin-b36730a3efc8-x86_64-apple-darwin-codesigning.tar.gz
1017504a58bc5c1ba457dcbea48572b336333693885c292cf644925513cfb393  guix-build-b36730a3efc8/output/x86_64-apple-darwin/bitcoin-b36730a3efc8-x86_64-apple-darwin-unsigned.tar.gz
d82720b34eb948716304084506906dffb5c54a2c5c875ac6fb029eaa5b67de2f  guix-build-b36730a3efc8/output/x86_64-apple-darwin/bitcoin-b36730a3efc8-x86_64-apple-darwin-unsigned.zip
06d242a85329e4ba73385fc6f43c75e766d9ec7cbd54f66d7a6431c4b6242ef4  guix-build-b36730a3efc8/output/x86_64-linux-gnu/SHA256SUMS.part
5156a48188c0e6a5448d67d335dcea0444b734e832446ec4cd9236f7ddd2feec  guix-build-b36730a3efc8/output/x86_64-linux-gnu/bitcoin-b36730a3efc8-x86_64-linux-gnu-debug.tar.gz
3319ee73fc87f87a554f29f09994e96ef51424be8e2535d00963ff9291761143  guix-build-b36730a3efc8/output/x86_64-linux-gnu/bitcoin-b36730a3efc8-x86_64-linux-gnu.tar.gz
724e29ad87d45682faf8a26e6ec70408a96d22c56ec21c04e128e57757ba4990  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/SHA256SUMS.part
bf6093cca7f4741b9a68d4e86766986c1c764cd05c93f23adec9bfd3641fdcde  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-codesigning.tar.gz
f32aead13c295e61774a00e289bab5deced84e184b5c8aae4559aaa5526e2ea1  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-debug.zip
3bdc7bd3175f4027f8e9c0ad140f4445a0ef6022d6c281a5df3540e59c7eda4a  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-setup-unsigned.exe
2a318a377ec0623a944cd5240e50677b4ea6bc2620f48edec9c143877f9ae85a  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-unsigned.zip
```

sedited commented:
Guix build (aarch64):

```
6be5510ac97201592d603a27c9a505de38c458114788d70dd661fe8fd3d6e6f8  guix-build-b36730a3efc8/output/aarch64-linux-gnu/SHA256SUMS.part
5e45521428d5238e033a7d389fff528cbca87ab90896064d750d025640703c69  guix-build-b36730a3efc8/output/aarch64-linux-gnu/bitcoin-b36730a3efc8-aarch64-linux-gnu-debug.tar.gz
f6b3984af843fe120e66e9666b4b8a089ce2b840b9a2357d8b44e2a8eda4410b  guix-build-b36730a3efc8/output/aarch64-linux-gnu/bitcoin-b36730a3efc8-aarch64-linux-gnu.tar.gz
4c8e7b5f32293878baa2182e64195ff0c0dfb17241d0b1f08e1d4f4a39f4adfd  guix-build-b36730a3efc8/output/arm-linux-gnueabihf/SHA256SUMS.part
44f106cccc1d7c140109476c94d60742e6cd7bae5add7524d22447f7cc9721de  guix-build-b36730a3efc8/output/arm-linux-gnueabihf/bitcoin-b36730a3efc8-arm-linux-gnueabihf-debug.tar.gz
77a191511717860710465402eaf76682b9febda71877faf4d808017dfe610825  guix-build-b36730a3efc8/output/arm-linux-gnueabihf/bitcoin-b36730a3efc8-arm-linux-gnueabihf.tar.gz
553c96e9f0b89fb1aa36a92f1b7358756afb18f74ef228bc48d843dcc74955f2  guix-build-b36730a3efc8/output/arm64-apple-darwin/SHA256SUMS.part
c6722e53e2297a0e4ba278f0a9c2db7f5aac46cd08a8c306dab9a2c717aed357  guix-build-b36730a3efc8/output/arm64-apple-darwin/bitcoin-b36730a3efc8-arm64-apple-darwin-codesigning.tar.gz
d812ae4608f5a82cf426a2092e368d0a65414e854c29124f49a86cf914eff94d  guix-build-b36730a3efc8/output/arm64-apple-darwin/bitcoin-b36730a3efc8-arm64-apple-darwin-unsigned.tar.gz
c5f54764eae63cfbf0eb029ba7c8c82073cf8bb9fd577bf780f701d09beccc1a  guix-build-b36730a3efc8/output/arm64-apple-darwin/bitcoin-b36730a3efc8-arm64-apple-darwin-unsigned.zip
1266fafab396a1fc3395ea4767c8e36225b05537f34ba759936677f07dced0f3  guix-build-b36730a3efc8/output/dist-archive/bitcoin-b36730a3efc8.tar.gz
2989ef62e3ea922f4d20b2cd819ea6d5a92fee9df44e0625aa3fee8de0126688  guix-build-b36730a3efc8/output/powerpc64-linux-gnu/SHA256SUMS.part
2839d7dd674212a55e3c4607c371233b11a7532e080c2f7356732aa629ecda7a  guix-build-b36730a3efc8/output/powerpc64-linux-gnu/bitcoin-b36730a3efc8-powerpc64-linux-gnu-debug.tar.gz
474fe25da36be086ea228a4d5d7684f49fdb0a6227bc10d49a6adb796921316d  guix-build-b36730a3efc8/output/powerpc64-linux-gnu/bitcoin-b36730a3efc8-powerpc64-linux-gnu.tar.gz
590191adf09b41365e80d28e9b1ebee7961154cb733767aad2ca5d38ad3073b8  guix-build-b36730a3efc8/output/riscv64-linux-gnu/SHA256SUMS.part
61dacc898e24b439d747840b3551b76731809529ae426caad9e0427eb86db641  guix-build-b36730a3efc8/output/riscv64-linux-gnu/bitcoin-b36730a3efc8-riscv64-linux-gnu-debug.tar.gz
af959fce01450358ec25da5400d1aaf004fa43a7ba5795833036beeb6ee0a935  guix-build-b36730a3efc8/output/riscv64-linux-gnu/bitcoin-b36730a3efc8-riscv64-linux-gnu.tar.gz
31925acd25f9e26e29ecdafdf1b4866e457aa08d64154aa5a82bbca1e49beaba  guix-build-b36730a3efc8/output/x86_64-apple-darwin/SHA256SUMS.part
b6c36c78dac4145dc37e2318913b778e26a05e84e3549632daaa28dfc11e270c  guix-build-b36730a3efc8/output/x86_64-apple-darwin/bitcoin-b36730a3efc8-x86_64-apple-darwin-codesigning.tar.gz
1017504a58bc5c1ba457dcbea48572b336333693885c292cf644925513cfb393  guix-build-b36730a3efc8/output/x86_64-apple-darwin/bitcoin-b36730a3efc8-x86_64-apple-darwin-unsigned.tar.gz
d82720b34eb948716304084506906dffb5c54a2c5c875ac6fb029eaa5b67de2f  guix-build-b36730a3efc8/output/x86_64-apple-darwin/bitcoin-b36730a3efc8-x86_64-apple-darwin-unsigned.zip
06d242a85329e4ba73385fc6f43c75e766d9ec7cbd54f66d7a6431c4b6242ef4  guix-build-b36730a3efc8/output/x86_64-linux-gnu/SHA256SUMS.part
5156a48188c0e6a5448d67d335dcea0444b734e832446ec4cd9236f7ddd2feec  guix-build-b36730a3efc8/output/x86_64-linux-gnu/bitcoin-b36730a3efc8-x86_64-linux-gnu-debug.tar.gz
3319ee73fc87f87a554f29f09994e96ef51424be8e2535d00963ff9291761143  guix-build-b36730a3efc8/output/x86_64-linux-gnu/bitcoin-b36730a3efc8-x86_64-linux-gnu.tar.gz
724e29ad87d45682faf8a26e6ec70408a96d22c56ec21c04e128e57757ba4990  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/SHA256SUMS.part
bf6093cca7f4741b9a68d4e86766986c1c764cd05c93f23adec9bfd3641fdcde  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-codesigning.tar.gz
f32aead13c295e61774a00e289bab5deced84e184b5c8aae4559aaa5526e2ea1  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-debug.zip
3bdc7bd3175f4027f8e9c0ad140f4445a0ef6022d6c281a5df3540e59c7eda4a  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-setup-unsigned.exe
2a318a377ec0623a944cd5240e50677b4ea6bc2620f48edec9c143877f9ae85a  guix-build-b36730a3efc8/output/x86_64-w64-mingw32/bitcoin-b36730a3efc8-win64-unsigned.zip

```

fanquake commented:
ACK b36730a3efc85c4acdb93f06910e37151b8f2cf9
